File: fc-chat/demo/part3/ch3/02-5-message-system-add-connection/src/main/java/net/prostars/messagesystem/repository/UserRepository.java
package net.prostars.messagesystem.repository;

import jakarta.persistence.LockModeType;
import java.util.Optional;
import net.prostars.messagesystem.dto.projection.CountProjection;
import net.prostars.messagesystem.dto.projection.InviteCodeProjection;
import net.prostars.messagesystem.dto.projection.UsernameProjection;
import net.prostars.messagesystem.entity.UserEntity;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.data.jpa.repository.Lock;
import org.springframework.lang.NonNull;
import org.springframework.stereotype.Repository;

@Repository
public interface UserRepository extends JpaRepository<UserEntity, Long> {

  Optional<UserEntity> findByUsername(@NonNull String username);

  Optional<UsernameProjection> findByUserId(@NonNull Long userId);

  Optional<UserEntity> findByConnectionInviteCode(@NonNull String connectionInviteCode);

  Optional<InviteCodeProjection> findInviteCodeByUserId(@NonNull Long userId);

  Optional<CountProjection> findCountByUserId(@NonNull Long userId);

  @Lock(LockModeType.PESSIMISTIC_WRITE)
  Optional<UserEntity> findForUpdateByUserId(@NonNull Long userId);
}

File: fc-chat/demo/part3/with-push/src/main/kotlin/com/yscorp/withpush/messagesystem/repository/UserRepository.kt
package com.yscorp.withpush.messagesystem.repository

import com.yscorp.withpush.messagesystem.dto.projection.CountProjection
import com.yscorp.withpush.messagesystem.dto.projection.InviteCodeProjection
import com.yscorp.withpush.messagesystem.dto.projection.UserIdProjection
import com.yscorp.withpush.messagesystem.dto.projection.UsernameProjection
import com.yscorp.withpush.messagesystem.entity.UserEntity
import jakarta.persistence.LockModeType
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.data.jpa.repository.Lock
import org.springframework.stereotype.Repository

@Repository
interface UserRepository : JpaRepository<UserEntity, Long> {

    fun findByUsername(username: String): UserEntity?

    fun findByUsernameIn(usernames: Collection<String>): List<UserIdProjection>

    fun findByUserId(userId: Long): UsernameProjection?

    fun findByInviteCode(inviteCode: String): UserEntity?

    fun findInviteCodeByUserId(userId: Long): InviteCodeProjection?

    fun findCountByUserId(userId: Long): CountProjection?

    @Lock(LockModeType.PESSIMISTIC_WRITE)
    fun findForUpdateByUserId(userId: Long): UserEntity?
}
